tests/wol/test_wol.py

```diff
@@ -0,0 +1,387 @@
+import binascii
+import logging
+import pytest
+import random
+import tempfile
+import time
+from socket import inet_aton
+from scapy.all import sniff as scapy_sniff
+from tests.common.utilities import capture_and_check_packet_on_dut
+from tests.common.helpers.assertions import pytest_assert
+
+pytestmark = [
+    pytest.mark.topology('mx'),
+]
+
+WOL_SLL_PKT_FILTER = 'ether[14:2]==0x0842'
+WOL_ETHER_PKT_FILTER = 'ether[12:2]==0x0842'
+BROADCAST_MAC = 'ff:ff:ff:ff:ff:ff'
+ETHER_TYPE_WOL_BIN = b'\x08\x42'
+ETHER_TYPE_WOL_DEC = int('842', 16)
+PACKET_TYPE_BROADCAST = 1
+PACKET_TYPE_UNICAST = 3
+LINK_LAYER_TYPE_ETHER = 1
+VLAN_MEMBER_CHANGE_ERR = r'.*Failed to get port by bridge port ID .*'
+
+
+def generate_pcap_file_path(id: str) -> str:
+    return '/tmp/wol_test_%s.pcap' % id
+
+
+def vlan_n2i(vlan_name):
+    """
+        Convert vlan name to vlan id
+    """
+    return vlan_name.replace("Vlan", "")
+
+
+def p2b(password: str) -> bytes:
+    """
+        convert password to bytes
+    """
+    if not password:
+        return b''
+    if ':' in password:
+        return binascii.unhexlify(password.replace(':', ''))
+    if '.' in password:
+        return inet_aton(password)
+    pytest.fail("invalid password %s" % password)
+
+
+def m2b(mac: str) -> bytes:
+    """
+        convert mac address to bytes
+    """
+    return binascii.unhexlify(mac.replace(':', ''))
+
+
+def build_magic_packet(src_mac: str, target_mac: str, broadcast: bool, password: str = "") -> bytes:
+    dst_mac = BROADCAST_MAC if broadcast else target_mac
+    return m2b(dst_mac) + m2b(src_mac) + ETHER_TYPE_WOL_BIN \
+        + build_magic_packet_payload(target_mac, password)
+
+
+def build_magic_packet_payload(target_mac: str, password: str = "") -> bytes:
+    return b'\xff' * 6 + m2b(target_mac) * 16 + p2b(password)
+
+
+def test_send_to_single_specific_interface(
+    duthost,
+    ptfhost,
+    get_connected_dut_intf_to_ptf_index
+):
+    dut_mac = duthost.facts['router_mac']
+    target_mac = "1a:2b:3c:d1:e2:f0"
+    connected_dut_intf_to_ptf_index = get_connected_dut_intf_to_ptf_index
+    random_dut_port, randon_ptf_port = random.choice(connected_dut_intf_to_ptf_index)
+    logging.info("Test with random dut port %s and ptf port index %s" % (random_dut_port, randon_ptf_port))
+
+    def validate_wol_broadcast_port(pkts):
```

Comment:
Yes, It's possible. 

But I prefer the current version, because it's easy to read and edit with closures feature in python.

If we reuse this function more times in future, we can make this function standalone.